src/main/scala/org/broadinstitute/dsde/firecloud/EntityClient.scala

```diff
@@ -340,4 +348,49 @@ class EntityClient (requestContext: RequestContext, modelSchema: ModelSchema)(im
       }
     }
   }
+
+  def importPFB(pipeline: WithTransformerConcatenation[HttpRequest, Future[HttpResponse]],
+                workspaceNamespace: String, workspaceName: String, pfbRequest: PfbImportRequest): Future[PerRequestMessage] = {
+    def callArrow = {
+      logger.debug("calling arrow")
+      pipeline {
+        Post(avroToRawlsURL, requestContext.request.entity)
+      }
+    }
+
+    def callRawls(arrowResponse: HttpResponse) = {
+      logger.debug("calling rawls")
+      pipeline {
+        Post(FireCloudDirectiveUtils.encodeUri(Rawls.entityPathFromWorkspace(workspaceNamespace, workspaceName) + "/batchUpsert"), arrowResponse.entity)
+      }
+    }
+
+    val pfbUrl = new URL(pfbRequest.url.replace(" ", "%20"))
+    val acceptableProtocols = Seq("https") //for when we inevitably change our mind and need to support others
+
+    if (!acceptableProtocols.contains(pfbUrl.getProtocol)) {
+      Future.successful(RequestCompleteWithErrorReport(StatusCodes.BadRequest, "Invalid URL protocol: must be https only"))
+    } else {
+      val futureResponse = for {
+        arrowResponse <- callArrow
+        rawlsResponse <- arrowResponse.status match {
+          case OK =>
+            callRawls(arrowResponse)
+          case _ =>
+            Future.successful(arrowResponse)
+        }
+      } yield (rawlsResponse.status match {
+        case NoContent =>
```

Comment:
i found some of this naming confusing. like the rawlsResponse might not actually be from rawls. and for some reason i associated the Future.succesful(arrowResponse) with the futureResponse val (instead of the rawlsResponse) when i first looked at the code. i don't really have any good suggestions though... maybe i've just been away from scala too long!

Reply:
I wrestled with this too. I think unrolling the for-comprehension avoided that particular naming issue, but overall I preferred the expressiveness of the for-comprehension.

`futureResponse` could be `futurePerRequestMessage` or `futureResult` or simply `future`. It represents the output of the function and only exists to express that the `Future.failed` case is handled using `recover`. Do any of those names seem better to you @ahaessly ?

Reply:
i was thinking that instead of rawlsResponse, it could be something like rawlsResonseOrArrowError (although that is pretty long). anything that indicates it may be the response from the rawls or arrow call. 

as for the futureResponse val, maybe don't assign the for result to a val but wrap the for in () and then just call recover after the closing paren. unless you find that more unreadable. if you stick with using the val, i think i like futurePerRequestMessage best.

      (for {
        arrowResponse <- callArrow
        ...
      })) recover {
        ...

Reply:
My scala was definitely rusty on the first pass. It was totally weird the way I was passing the arrow error through as a "success". The whole point of mapping over a sequence of futures using a for-comprehension is to take advantage of the success bias and skip remaining steps once there's a failure. I just got hung up on some type details.

I've refactored to split those out and embraced futures and I think it reads a bit better now.